### DISCUSSION TITLE
WeAct ePaper and PicoW

### DISCUSSION THREAD

**Discussion by hoibatter** [micropython/micropython#12416] [RP2040 / Pico]
Hi all,
I'm currently trying to get a PicoW and a WeAct 2.9" ePaper display working.
I'm pretty new to micropython and the Pico, so any help would be appreciated.

I'm not even sure if the wiring is correct, since the driver and the connectors on the display do not correspond..

The driver I've tried: https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/epaper/epd29_ssd1680.py
(The driver seems to be ok, since I'm not getting any errors)
But I can't seem to get any demo working. If you need me to provide more specific informations, let me know and I'll try to update the post asap.

Thanks and best regards

**Comment by domi-a** [comment]
are u having trouble with choosing the pins on the pico?
[pico pinout](https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html)

or are unsure about the side on the display to the setup in code?
maybe [this example from an esp32](https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/ssd1680_esp32.py) would give u enough advice?

**Reply by hoibatter** [reply]
Thanks for your reply! I think I've got the pin layout figured out, and I was able to successfully use the display with an esp32.
I was just not able to make it work using the Pico and micropython (which was my main goal...)

So if anyone has any suggestions about this particular display driven by a Pico, it would really help...

**Reply by peterhinch** [reply]
If you've used it with ESP32 and my driver it should definitely work with the Pico: the driver is standard MicroPython code. The fault is almost certainly in the pin definitions and wiring. There are two common errors with the Pico:
 1. Confusing GPIO numbers with pin numbers.
 2. Making wrong assumptions about default pins for the SPI interface. Each SPI has two alternative pinouts. I always specify the pins in the constructor call. See my setup examples such as [this one](https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/epd29_pico.py)

**Reply by hoibatter** [reply]
@peterhinch I'm sorry I wasn't clear on this one. When I was testing the display using an esp32, I have been using the official examples of the producer from here: https://github.com/WeActStudio/WeActStudio.EpaperModule/tree/master/Example

But anyways, I'll try again using your setup example and report back. Could be a couple of days though until I get my hands on this again.

**Reply by QC20** [reply]
Did you have any luck getting it to work? :-) 
- if so, please share.

**Reply by flaszlo2000** [reply]
For the future readers coming from googling:

This works.
You will need the following on your device (in this case a Pico):
https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/epaper/epd29_ssd1680.py
https://github.com/peterhinch/micropython-nano-gui/blob/master/drivers/boolpalette.py

After that, you will need to create a color_setup.py created manually on the root folder (*not in lib*) with the following:
https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/epd29_pico.py

its important to follow the _pico.py's idea because if you are going to be that bold like I was to use this instead:
https://github.com/peterhinch/micropython-nano-gui/blob/master/setup_examples/epd29_pyb_sync.py
Then your SCL (and other things) wont be set and only garbage will be drawn on your screen.

I was using an aliexpress pico and aliexpress WeAct 2.9 epaper display and managed to make it work.

**Reply by de-dh** [reply]
I've uploaded a complete nano-gui example here: [WeAct-2.9-EPD-Esp32-MPY](https://github.com/de-dh/WeAct-2.9-EPD-Esp32-MPY).
Both, the Esp32 and the Display were purchased from AliExpress.
You just need to upload the files to your Pico and configure the Pins in `color_setup.py`.
